### ISSUE TITLE
Purchase Confirmation JSON is valid?

### ISSUE THREAD

**Issue by nin9creative** [Simbul/baker#577]
I was doing some testing in trying to set up an API to receive the post from the Purchase Confirmation...

I'm not sure if this is correct.  This is the data that was captured in XCode as what was being posted.

However I think that the = is invalid JSON syntax according to JSONLINT

Parse error on line 2:
{    "receipt-data"=ewoJInNpZ25hdHVyZSI
-------------------^
Expecting '}', ':', ',', ']'

{
    "receipt-data" = ewoJInNpZ25hdHVyZSIgPSAiQW1zMWM2bkFLS2l5SndVM3MvcE9hYkdqYVhpbW0yMG5hcVlsRVd5WnJVWXJwZHBwN2xwa0FIN2Q3OVE4NFV5QWJCWG9ncnlERnVnRnlXNXdjbWl2NkhpMUZCOTk4ekV3aHFId2RsWUp1RGU4RDEzS3FrWmRCVklJbzByVkpZdjhKNGRhMGlKLzhVRjBJZ3JucEpHNjhBa0NIcGFVS3Q0VUMyVVNUTmFJOXUrTkFBQURWekNDQTFNd2dnSTdvQU1DQVFJQ0NHVVVrVTNaV0FTMU1BMEdDU3FHU0liM0RRRUJCUVVBTUg4eEN6QUpCZ05WQkFZVEFsVlRNUk13RVFZRFZRUUtEQXBCY0hCc1pTQkpibU11TVNZd0pBWURWUVFMREIxQmNIQnNaU0JEWlhKMGFXWnBZMkYwYVc5dUlFRjFkR2h2Y21sMGVURXpNREVHQTFVRUF3d3FRWEJ3YkdVZ2FWUjFibVZ6SUZOMGIzSmxJRU5sY25ScFptbGpZWFJwYjI0Z1FYVjBhRzl5YVhSNU1CNFhEVEE1TURZeE5USXlNRFUxTmxvWERURTBNRFl4TkRJeU1EVTFObG93WkRFak1DRUdBMVVFQXd3YVVIVnlZMmhoYzJWU1pXTmxhWEIwUTJWeWRHbG1hV05oZEdVeEd6QVpCZ05WQkFzTUVrRndjR3hsSUdsVWRXNWxjeUJUZEc5eVpURVRNQkVHQTFVRUNnd0tRWEJ3YkdVZ1NXNWpMakVMTUFrR0ExVUVCaE1DVlZNd2daOHdEUVlKS29aSWh2Y05BUUVCQlFBRGdZMEFNSUdKQW9HQkFNclJqRjJjdDRJclNkaVRDaGFJMGc4cHd2L2NtSHM4cC9Sd1YvcnQvOTFYS1ZoTmw0WElCaW1LalFRTmZnSHNEczZ5anUrK0RyS0pFN3VLc3BoTWRkS1lmRkU1ckdYc0FkQkVqQndSSXhleFRldngzSExFRkdBdDFtb0t4NTA5ZGh4dGlJZERnSnYyWWFWczQ5QjB1SnZOZHk2U01xTk5MSHNETHpEUzlvWkhBZ01CQUFHamNqQndNQXdHQTFVZEV3RUIvd1FDTUFBd0h3WURWUjBqQkJnd0ZvQVVOaDNvNHAyQzBnRVl0VEpyRHRkREM1RllRem93RGdZRFZSMFBBUUgvQkFRREFnZUFNQjBHQTFVZERnUVdCQlNwZzRQeUdVakZQaEpYQ0JUTXphTittVjhrOVRBUUJnb3Foa2lHOTJOa0JnVUJCQUlGQURBTkJna3Foa2lHOXcwQkFRVUZBQU9DQVFFQUVhU2JQanRtTjRDL0lCM1FFcEszMlJ4YWNDRFhkVlhBZVZSZVM1RmFaeGMrdDg4cFFQOTNCaUF4dmRXLzNlVFNNR1k1RmJlQVlMM2V0cVA1Z204d3JGb2pYMGlreVZSU3RRKy9BUTBLRWp0cUIwN2tMczlRVWU4Y3pSOFVHZmRNMUV1bVYvVWd2RGQ0TndOWXhMUU1nNFdUUWZna1FRVnk4R1had1ZIZ2JFL1VDNlk3MDUzcEdYQms1MU5QTTN3b3hoZDNnU1JMdlhqK2xvSHNTdGNURXFlOXBCRHBtRzUrc2s0dHcrR0szR01lRU41LytlMVFUOW5wL0tsMW5qK2FCdzdDMHhzeTBiRm5hQWQxY1NTNnhkb3J5L0NVdk02Z3RLc21uT09kcVRlc2JwMGJzOHNuNldxczBDOWRnY3hSSHVPTVoydG04bnBMVW03YXJnT1N6UT09IjsKCSJwdXJjaGFzZS1pbmZvIiA9ICJld29KSW05eWFXZHBibUZzTFhCMWNtTm9ZWE5sTFdSaGRHVXRjSE4wSWlBOUlDSXlNREV5TFRFeUxUQTFJREl5T2pJME9qUTJJRUZ0WlhKcFkyRXZURzl6WDBGdVoyVnNaWE1pT3dvSkluVnVhWEYxWlMxcFpHVnVkR2xtYVdWeUlpQTlJQ0l3TURBd1lqQXpNV001WmpnaU93b0pJbTl5YVdkcGJtRnNMWFJ5WVc1ellXTjBhVzl1TFdsa0lpQTlJQ0l4TURBd01EQXdNRFU1T1RNNE1USTBJanNLQ1NKaWRuSnpJaUE5SUNKQ1lXdGxjaTAwTUNJN0Nna2lkSEpoYm5OaFkzUnBiMjR0YVdRaUlEMGdJakV3TURBd01EQXdOVGs1TXpneE1qUWlPd29KSW5GMVlXNTBhWFI1SWlBOUlDSXhJanNLQ1NKdmNtbG5hVzVoYkMxd2RYSmphR0Z6WlMxa1lYUmxMVzF6SWlBOUlDSXhNelUwTnpjMU1EZzJPRFV3SWpzS0NTSndjbTlrZFdOMExXbGtJaUE5SUNKamIyMHVibWx1T1dOeVpXRjBhWFpsTG0xaFozSnZZMnRsZEM1emRXSXVabkpsWlNJN0Nna2lhWFJsYlMxcFpDSWdQU0FpTlRnMU1UTXpOVFF4SWpzS0NTSmlhV1FpSUQwZ0ltTnZiUzV1YVc0NVkzSmxZWFJwZG1VdWJXRm5jbTlqYTJWMElqc0tDU0p3ZFhKamFHRnpaUzFrWVhSbExXMXpJaUE5SUNJeE16VTBOemMxTURnMk9EVXdJanNLQ1NKd2RYSmphR0Z6WlMxa1lYUmxJaUE5SUNJeU1ERXlMVEV5TFRBMklEQTJPakkwT2pRMklFVjBZeTlIVFZRaU93b0pJbkIxY21Ob1lYTmxMV1JoZEdVdGNITjBJaUE5SUNJeU1ERXlMVEV5TFRBMUlESXlPakkwT2pRMklFRnRaWEpwWTJFdlRHOXpYMEZ1WjJWc1pYTWlPd29KSW05eWFXZHBibUZzTFhCMWNtTm9ZWE5sTFdSaGRHVWlJRDBnSWpJd01USXRNVEl0TURZZ01EWTZNalE2TkRZZ1JYUmpMMGROVkNJN0NuMD0iOwoJImVudmlyb25tZW50IiA9ICJTYW5kYm94IjsKCSJwb2QiID0gIjEwMCI7Cgkic2lnbmluZy1zdGF0dXMiID0gIjAiOwp9;
}

**Comment by Simbul** [comment]
That's an interesting point. That JSON is generated by JSONKit, but the `=` definitely looks out of place.
We'll look into it.

**Comment by nin9creative** [comment]
I was able to fix this by making some tweaks.  However the code that I wrote actually spins off an asynchronous call because in my version of Baker I care about the result that comes back from the call, so I had to make more changes.

However this is what my version looks like (you'd need to change the NSURL connection so that it doesn't expect the delegate for the Baker master.  I don't have time right now to put this in a Pull Request.  I also had to add a header for the content type because the API code on my end needs that to determine the request type.

```
-(void)recordTransaction:(SKPaymentTransaction *)transaction {
    [[NSUserDefaults standardUserDefaults] setObject:transaction.transactionIdentifier forKey:@"receipt"];

    if ([PURCHASE_CONFIRMATION_URL length] > 0) {
        NSString *receiptData = [transaction.transactionReceipt base64EncodedString];
        NSDictionary *jsonDict = [NSDictionary dictionaryWithObjectsAndKeys:
                                  receiptData, @"receipt-data",
                                  nil];
        NSError *error = nil;

        NSString *jsonRequest = [jsonDict JSONString];

        NSLog(@"jsonRequest is %@", jsonRequest);       

        NSData *requestData = [NSData dataWithBytes:[jsonRequest UTF8String] length:[jsonRequest length]];

        if (error) {
            NSLog(@"Error generating receipt JSON: %@", error);
        } else {
            NSURL *requestURL = [NSURL URLWithString:PURCHASE_CONFIRMATION_URL];
            NSMutableURLRequest *req = [[NSMutableURLRequest alloc] initWithURL:requestURL];

            [req setHTTPMethod:@"POST"];
            [req setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
            [req setHTTPBody:requestData];

            NSURLConnection *conn = [[NSURLConnection alloc] initWithRequest:req delegate:self];
            if (conn) {
                NSLog(@"Posting App Store transaction receipt to %@", PURCHASE_CONFIRMATION_URL);
            } else {
                NSLog(@"Cannot connect to %@", PURCHASE_CONFIRMATION_URL);
            }
        }

    }
}
```

**Comment by Simbul** [comment]
This should have been fixed in f1db74af05cc5251fad4736108c183b4a94109a1 (on the `paid-subscriptions` branch).
With that change, we won't be sending a plain JSON anymore, but a proper request with parameters. This should make it easier to parse it on the server side (a PHP example is on [baker-server-example](https://github.com/Simbul/baker-server-example)).

We'll keep this open until that change is merged back into `master`.

**Comment by Simbul** [comment]
Closing as the change has been merged into `master`.
